File: problem47.py
def trial_division(n, bound=None):
  if n == 1: return 1
  for p in [2, 3, 5]:
    if n%p == 0: return p
  if bound == None: bound = n
  dif = [6, 4, 2, 4, 2, 4, 6, 2]
  m = 7; i = 1
  while m <= bound and m*m <= n:
    if n%m == 0:
      return m
    m += dif[i%8]
    i += 1
  return n

def factor(n):
  if n in [-1, 0, 1]: return []
  if n < 0: n = -n
  F = []
  while n != 1:
    p = trial_division(n)
    e = 1
    n /= p
    while n%p == 0:
      e += 1; n /= p
    F.append((p,e))
  F.sort()
  return F
 
 
ci = 1
nf = 4		#number of distinct factors
ns = 4		#number of consecutive integers
n = 2*3*5*7	#starting candidate for search
while ci != ns:
  n += 1
  if len(factor(n)) == nf: ci += 1
  else: ci = 0

print "Answer to PE47 = ", n-nf+1

File: problem47.rb
require "mathn"
require "set"

# The first two consecutive numbers to have two distinct prime factors are:
#
# 14 = 2 × 7
# 15 = 3 × 5
# 
# The first three consecutive numbers to have three distinct prime factors are:
# 
# 644 = 2² × 7 × 23
# 645 = 3 × 5 × 43
# 646 = 2 × 17 × 19.
# 
# Find the first four consecutive integers to have four distinct prime factors. What is the first of these numbers?

def prime_factors(n)
  return [] if n == -1 || n == 0 || n == 1
  
  n = n.abs
  
  factors = []
  
  while n > 1
    prime = false
    Prime.each(n) do |p|
      if n % p == 0
        prime = true
        factors << p
        n /= p
        break
      end
    end
    break unless prime
  end
  
  factors
end

# lowest number with four prime numbers as factors is
num = 2 * 3 * 5 * 7 + 1

count = 0
while count < 4
  factors = prime_factors(num)
  
  if factors.uniq.length == 4
    puts "#{num}: #{factors}"
    count += 1
  else
    count = 0
  end
  
  num += 1
end

puts "Num: #{num-4}"



# answer: 134043


    
exit


primes = Prime.each(1000).to_set

#sorted = primes.to_a.combination(4).map { |a, b, c, d| a * b * c * d }.sort

products = []
primes.to_a.combination().each do |a, b, c|
  
end

puts "possibilities: #{sorted.length}"

#sorted = primes.to_a.sort

0.upto(sorted.length-4) do |i|
  if  sorted[i]+1 == sorted[i+1] &&
      sorted[i]+2 == sorted[i+2] #&&
      #sorted[i]+3 == sorted[i+3]
    puts "First: #{sorted[i]} (#{sorted[i+1]}, #{sorted[i+2]})"
    break
  end
end

exit

count = 0
sorted.each do |i|
  count += 1
  print "#{i}, "
  if count % 10 == 0
    gets
  end
end